replace should not decrement the len of the slab

rrichardson opened the pull request:
minor change to move the decrementing of length from replace_() into remove()

dpc commented:
I thin this should have some `#[test]` clearly showing the original problem and documenting designed behavior.

rrichardson commented:
I agree.  This is a good idea in principle, and especially applicable here, since there are a couple different routes to adding and removing items.

rrichardson commented:
OK.  I have updated and added replace() tests which showcase the length issue. This also alerted me to a case where we should update the length of the slab if we're replacing None with Some. 

I have the original codebase with the updated tests here: https://github.com/rrichardson/slab/tree/orig 

You'll notice both the replace and replace_again fail,  replace_again panics with an arithmetic underflow, and replace fails the appropriate count() check.
